modules/perception/obstacle/camera/tracker/cascaded_camera_tracker.cc

```diff
@@ -0,0 +1,117 @@
+/******************************************************************************
+ * Copyright 2018 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/perception/obstacle/camera/tracker/cascaded_camera_tracker.h"
+
+namespace apollo {
+namespace perception {
+
+bool CascadedCameraTracker::Init() {
+  bool init_flag = true;
+
+  init_flag &= cs2d_tracker_.Init();
+  if (dl_feature_) init_flag &= dlf_tracker_.Init();
+  init_flag &= kcf_tracker_.Init();
+
+  return init_flag;
+}
+
+bool CascadedCameraTracker::Associate(const cv::Mat& img,
+                                      const float& timestamp,
+                                      std::vector<VisualObjectPtr>* objects) {
+  frame_idx_++;
+
+  // detect id to (track id, first_timestamp) mapping
+  std::unordered_map<int, std::pair<int, float>> id_mapping;
+
+  float scale = 1.0f;
+  std::vector<Detected> detected;
+  GetDetectedFromVO(img.size(), scale, *objects, &detected);
+
+  // Affinity matrix
+  std::vector<std::vector<float>> affinity_matrix;
+  affinity_matrix = std::vector<std::vector<float>>(
+      tracks_.size(), std::vector<float>(detected.size(), 1.0f));
+
+  // cs2d
+  std::vector<std::vector<float>> cs2d_affinity_matrix;
+  cs2d_tracker_.SelectFull(tracks_.size(), detected.size());
+  cs2d_tracker_.GetAffinityMatrix(img, tracks_, detected,
+                                  &cs2d_affinity_matrix);
+  MergeAffinityMatrix(cs2d_affinity_matrix, &affinity_matrix);
+
+  // dlf
+  if (dl_feature_) {
+    std::vector<std::vector<float>> dlf_affinity_matrix;
+    dlf_tracker_.SelectFull(tracks_.size(), detected.size());
+    dlf_tracker_.GetAffinityMatrix(img, tracks_, detected,
+                                   &dlf_affinity_matrix);
+
+    // Merge
+    MergeAffinityMatrix(dlf_affinity_matrix, &affinity_matrix);
+
+    // High confidence selection filtering. Thresholds are tuned in 3 passes
+    // For entry higher than threshold, see if there is other competing
+    // high scores in the cross (same row or column)
+    FilterAffinityMatrix(0.95f, 0.95f, &affinity_matrix);
```

Comment:
may want to change this one to proto configs later for easier tuning.

Reply:
Yes, will do it later when tuning is needed